crates/control/src/services/builds_root.rs

```diff
@@ -0,0 +1,344 @@
+            ours.clone()
+        } else {
+            let build_state = LocalBuildState::default();
+            locals.insert(build_id.to_owned(), build_state.clone());
+            build_state
+        }
+    }
+}
+
+/// Creates builds root services from the `BuildsRootSettings`. Returns an error if the
+/// configuration is structurally invalid, but does not attempt to connect to any external
+/// services.
+pub fn init_builds_root(
+    conf: &config::BuildsRootSettings,
+) -> anyhow::Result<(PutBuilds, FetchBuilds)> {
+    if !conf.uri.path().ends_with('/') {
+        anyhow::bail!("invalid uri: '{}', must end with a '/'", conf.uri);
+    }
+
+    let root: Arc<dyn BuildsRootService> = match conf.uri.scheme() {
+        "gs" => Arc::new(gcs::GCSBuildsRoot::new(conf.uri.clone())?),
+        "file" => Arc::new(LocalBuildsRoot::new(conf.uri.path())),
+        other => anyhow::bail!("invalid uri: unsupported scheme: '{}'", other),
+    };
+    Ok((PutBuilds(root.clone()), FetchBuilds::new(root)))
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum BuildsRootError {
+    #[error("build database error: {0}")]
+    Sqlite(#[from] rusqlite::Error),
+    #[error("cloud storage error: {0}")]
+    GsUtil(#[from] SubprocessError),
+
+    #[error("cannot create build URI: {0}")]
+    Url(#[from] url::ParseError),
+    #[error("filesystem error: {0}")]
+    Io(#[from] std::io::Error),
+
+    #[error("{0}")]
+    PrevError(String),
+}
+
+#[async_trait]
+trait BuildsRootService: Debug + Send + Sync {
+    async fn put_build(&self, build_id: &str, build: &Path) -> Result<(), BuildsRootError>;
+    async fn retrieve_build(&self, build_id: &str) -> Result<PathBuf, BuildsRootError>;
+}
+
+// A `BuildsRootService` that stores all builds in a local directory.
+#[derive(Debug)]
+struct LocalBuildsRoot {
+    dir: PathBuf,
+}
+
+impl LocalBuildsRoot {
+    fn new(dir: impl AsRef<Path>) -> LocalBuildsRoot {
+        LocalBuildsRoot {
+            dir: dir.as_ref().to_owned(),
+        }
+    }
+}
+
+#[async_trait]
+impl BuildsRootService for LocalBuildsRoot {
+    async fn put_build(&self, build_id: &str, build: &Path) -> Result<(), BuildsRootError> {
+        use std::io;
+
+        let dest_path = self.dir.join(build_id);
+        if dest_path.exists() {
+            Err(BuildsRootError::Io(io::Error::new(
+                io::ErrorKind::Other,
+                format!("the build file: '{}' already exists", dest_path.display()),
+            )))
+        } else {
+            tokio::fs::copy(build, &dest_path).await?;
+            Ok(())
+        }
+    }
+
+    async fn retrieve_build(&self, build_id: &str) -> Result<PathBuf, BuildsRootError> {
+        use std::io;
+
+        let dest_path = self.dir.join(build_id);
+        if dest_path.exists() {
+            Ok(dest_path)
+        } else {
+            Err(BuildsRootError::Io(io::Error::new(
+                io::ErrorKind::NotFound,
+                "no such build exists within the root",
+            )))
```

Comment:
Whether a build is missing or not is a condition that has to be checked in the database. The GCS impl can't return a typed error saying that a build is missing, and so it seemed best not to imply that it could be having a `BuildsRootError::Missing` variant.

Reply:
No, that was the problem that I'd tried to bring up in conversation with you and Johnny. My understanding, based on that conversation, is that we're relying on the database to tell us whether or not a build _should_ exist, and using that to render a response if it's missing. So if the builds root service returns an error due to a missing build, we'd consider it an "internal" error, as it represents an inconsistency between the database a cloud storage.